### PR TITLE
gh-2027 ToByteSerialisers equals hashcode

### PR DESCRIPTION
**Pull request by c015dariu**
#2027 

Implemented equals using the given equals function and additionally the equals builder on serialisers which had non-static fields. Hashcode similarly implemented.

### CODE REVIEW

File: core/serialisation/src/test/java/uk/gov/gchq/gaffer/serialisation/ToBytesSerialisationTest.java [hunk 1]
```diff
@@ -23,8 +23,7 @@
 
 import java.util.Arrays;
 
-import static org.junit.Assert.assertArrayEquals;
-import static org.junit.Assert.assertNull;
+import static org.junit.Assert.*;
```
Comment (m55624) [inline core/serialisation/src/test/java/uk/gov/gchq/gaffer/serialisation/ToBytesSerialisationTest.java]: Can we keep these individual please

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/schema/SchemaTest.java [hunk 2]
```diff
@@ -47,23 +47,9 @@
...
-import static org.junit.Assert.assertArrayEquals;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertNotSame;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertSame;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
+import java.util.*;
```
Comment (m55624) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/schema/SchemaTest.java]: Same with these please

File: core/serialisation/src/test/java/uk/gov/gchq/gaffer/serialisation/ToBytesSerialisationTest.java (hunk 1 above)
Comment (c015dariu) [reply core/serialisation/src/test/java/uk/gov/gchq/gaffer/serialisation/ToBytesSerialisationTest.java]: Fixed as of last commit.

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/schema/SchemaTest.java (hunk 2 above)
Comment (c015dariu) [reply core/store/src/test/java/uk/gov/gchq/gaffer/store/schema/SchemaTest.java]: Fixed as of last commit.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirItemsSketchSerialiser.java [hunk 3]
```diff
@@ -62,4 +64,23 @@ public boolean preservesObjectOrdering() {
     public boolean isConsistent() {
         return false;
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
```
Comment (m55624) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirItemsSketchSerialiser.java]: Can we keep these to a standard form please, they are just easier to read if they all read the same, something like this:
``` @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }

        if (null == obj || getClass() != obj.getClass()) {
            return false;
        }

        final ReservoirItemsSketchSerialiser serialiser = (ReservoirItemsSketchSerialiser) obj;

        return new EqualsBuilder()
                .append(arrayOfItemsSerDe, serialiser.arrayOfItemsSerDe)
                .isEquals();
    }```

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/dataframe/MyPropertySerialiser.java [hunk 4]
```diff
@@ -59,4 +62,23 @@ public boolean preservesObjectOrdering() {
     public boolean isConsistent() {
         return false;
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/dataframe/MyPropertySerialiser.java]: Same here

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/testutils/MyPropertySerialiser.java [hunk 5]
```diff
@@ -60,4 +62,23 @@ public boolean preservesObjectOrdering() {
     public boolean isConsistent() {
         return false;
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
+            final MyPropertySerialiser that = (MyPropertySerialiser) o;
```
Comment (m55624) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/testutils/MyPropertySerialiser.java]: And here.

File: library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/serialisation/kryo/WrappedKryoSerializer.java [hunk 6]
```diff
@@ -65,4 +67,23 @@ public T read(final Kryo kryo, final Input input, final Class<T> type) {
                     + " to a byte array", e);
         }
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
+            final WrappedKryoSerializer that = (WrappedKryoSerializer) o;
```
Comment (m55624) [inline library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/serialisation/kryo/WrappedKryoSerializer.java]: And here.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MapSerialiser.java [hunk 7]
```diff
@@ -186,19 +186,11 @@ public boolean equals(final Object o) {
         boolean rtn = (this == o);
         if (!rtn && o != null && getClass() == o.getClass()) {
             final MapSerialiser that = (MapSerialiser) o;
-            try {
-                rtn = new EqualsBuilder()
-                        /*
-                         * JSONSerialiser is used because ToByteSerialisers.equals checks for sames instance.
-                         * All these serialises would require equals method to be updated.
-                         */
-                        .append(JSONSerialiser.serialise(keySerialiser), JSONSerialiser.serialise(that.keySerialiser))
```
Comment (d47853) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MapSerialiser.java]: The import for JsonSerialiser needs to be removed

File: core/type/src/main/java/uk/gov/gchq/gaffer/serialisation/FreqMapSerialiser.java [hunk 8]
```diff
@@ -132,4 +135,23 @@ public boolean isConsistent() {
     public FreqMap deserialiseEmpty() {
         return new FreqMap();
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/type/src/main/java/uk/gov/gchq/gaffer/serialisation/FreqMapSerialiser.java]: See below comment.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/PropertiesSerialiser.java [hunk 9]
```diff
@@ -76,5 +79,23 @@ protected void deserialiseProperties(final byte[] bytes, final Properties proper
             properties.put(propertyName, property);
         }
     }
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/PropertiesSerialiser.java]: See below comment.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/EntityIdSerialiser.java [hunk 10]
```diff
@@ -87,4 +90,23 @@ public boolean preservesObjectOrdering() {
     public boolean isConsistent() {
         return null != vertexSerialiser && vertexSerialiser.isConsistent();
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
```
Comment (m55624) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/EntityIdSerialiser.java]: See below comment.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/ElementSerialiser.java [hunk 11]
```diff
@@ -88,4 +91,25 @@ public Element deserialiseEmpty() throws SerialisationException {
     public boolean isConsistent() {
         return false;
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
```
Comment (m55624) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/ElementSerialiser.java]: See below comment.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/ElementIdSerialiser.java [hunk 12]
```diff
@@ -115,4 +117,25 @@ public boolean isConsistent() {
         return null != edgeIdSerialiser && null != entityIdSerialiser
                 && edgeIdSerialiser.isConsistent() && entityIdSerialiser.isConsistent();
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
+            final ElementIdSerialiser that = (ElementIdSerialiser) o;
```
Comment (m55624) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/ElementIdSerialiser.java]: See below comment.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/EdgeIdSerialiser.java [hunk 13]
```diff
@@ -99,4 +102,25 @@ public boolean preservesObjectOrdering() {
     public boolean isConsistent() {
         return null != vertexSerialiser && vertexSerialiser.isConsistent();
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/EdgeIdSerialiser.java]: See below comment.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/SetSerialiser.java [hunk 14]
```diff
@@ -152,4 +155,25 @@ public void setSetClassString(final String classType) throws ClassNotFoundExcept
     public void setSetClass(final Class<? extends Set> setClass) {
         this.setClass = setClass;
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/SetSerialiser.java]: See below comment.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MultiSerialiserStorage.java [hunk 15]
```diff
@@ -145,6 +148,33 @@ private void clear() {
         classToKey.clear();
     }
 
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MultiSerialiserStorage.java]: See below comment.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MultiSerialiser.java [hunk 16]
```diff
@@ -153,9 +156,27 @@ public boolean isConsistent() {
...
-
     @Override
     public boolean canHandle(final Class clazz) {
         return supportedSerialisers.canHandle(clazz);
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MultiSerialiser.java]: See below comment.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MapSerialiser.java [hunk 17]
```diff
@@ -186,19 +186,11 @@ public boolean equals(final Object o) {
         boolean rtn = (this == o);
         if (!rtn && o != null && getClass() == o.getClass()) {
             final MapSerialiser that = (MapSerialiser) o;
```
Comment (m55624) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/implementation/MapSerialiser.java]: See below comment.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/ToBytesViaStringDeserialiser.java [hunk 18]
```diff
@@ -86,4 +88,22 @@ public final T deserialise(final byte[] allBytes, final int offset, final int le
 
     protected abstract String serialiseToString(final T object) throws SerialisationException;
 
+    @Override
+    public boolean equals(final Object o) {
+        boolean rtn = (this == o);
+        if (!rtn && o != null && getClass() == o.getClass()) {
```
Comment (m55624) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/ToBytesViaStringDeserialiser.java]: See below comment.

File: core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/ToBytesViaStringDeserialiser.java [hunk 19]
```diff
@@ -15,6 +15,8 @@
  */
 package uk.gov.gchq.gaffer.serialisation;
 
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
```
Comment (d47853) [inline core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/ToBytesViaStringDeserialiser.java]: Move these imports below the JsonIgnore import to fix checkstyle

**Comment by c015dariu** [comment]
I'm not sure why the build is failing one of the serialisation/deserialisation tests, I haven't touched the CustomMapSerialiser class at all except by adding a equals/hashcode function and yet it is failing when trying to serialise and deserialise something. I've tried to debug the class but I'm unsure of what it's doing when serialising. Any ideas? Current travis build has the errors.

**Comment by GCHQDev404** [comment]
The failure is in `CustomMapSerialiserTest>SerialisationTest.shouldSerialiseWithHistoricValues`.
Some tests check that an object can serialise and deserialise, but this test is used to ensure that the serialisation does not change compared to a historic values such as the hardcoded string of bytes in the test or data written to disc. This test highlights breaking change to serialising data to/from storage, that data will not be able to read in the future.

**Comment by GCHQDev404** [comment]
Basically the CustomMapSerialiser is no longer serialising the same as it was.

**Comment by c015dariu** [comment]
Seems I accidentally gave ToByteViaStringDeserialiser an overriding equals/hashcode, reverting that file to develop fixed all tests. Additionally DateSerialiser (deprecated) was missing it's equals/hashcode which was covered up by ToByteViaStringDeserialiser's implemented eq/hc. All tests passing now.

**Pull request merged**